cot/src/project.rs

```diff
@@ -87,605 +155,1439 @@ pub trait CotApp: Send + Sync {
+    // Bootstrapper types
+    /// The type of the request handler.
+    type RequestHandler: Debug;
+
+    // App context types
+    /// The type of the configuration.
+    type Config: Debug;
+    /// The type of the apps.
+    type Apps;
+    /// The type of the router.
+    type Router: Debug;
+    /// The type of the auth backend.
+    type AuthBackend;
+    /// The type of the database.
+    #[cfg(feature = "db")]
+    type Database: Debug;
+}
+
+/// First phase of bootstrapping a Cot project, the uninitialized phase.
+///
+/// # See also
+///
+/// See the details about the different bootstrap phases in the
+/// [`BootstrapPhase`] trait documentation.
+#[derive(Debug, Copy, Clone, PartialEq, Eq, Hash)]
+pub struct Uninitialized;
+
+impl sealed::Sealed for Uninitialized {}
+impl BootstrapPhase for Uninitialized {
+    type RequestHandler = ();
+    type Config = ();
+    type Apps = ();
+    type Router = ();
+    type AuthBackend = ();
+    #[cfg(feature = "db")]
+    type Database = ();
+}
+
+/// Second phase of bootstrapping a Cot project, the with-config phase.
+///
+/// # See also
+///
+/// See the details about the different bootstrap phases in the
+/// [`BootstrapPhase`] trait documentation.
+#[derive(Debug, Copy, Clone, PartialEq, Eq, Hash)]
+pub struct WithConfig;
+
+impl sealed::Sealed for WithConfig {}
+impl BootstrapPhase for WithConfig {
+    type RequestHandler = ();
+    type Config = Arc<ProjectConfig>;
+    type Apps = ();
+    type Router = ();
+    type AuthBackend = ();
+    #[cfg(feature = "db")]
+    type Database = ();
+}
+
+/// Third phase of bootstrapping a Cot project, the with-apps phase.
+///
+/// # See also
+///
+/// See the details about the different bootstrap phases in the
+/// [`BootstrapPhase`] trait documentation.
+#[derive(Debug, Copy, Clone, PartialEq, Eq, Hash)]
+pub struct WithApps;
+
+impl sealed::Sealed for WithApps {}
+impl BootstrapPhase for WithApps {
+    type RequestHandler = ();
+    type Config = <WithConfig as BootstrapPhase>::Config;
+    type Apps = Vec<Box<dyn App>>;
+    type Router = Arc<Router>;
+    type AuthBackend = ();
+    #[cfg(feature = "db")]
+    type Database = ();
+}
+
+/// The final phase of bootstrapping a Cot project, the initialized phase.
+///
+/// # See also
+///
+/// See the details about the different bootstrap phases in the
+/// [`BootstrapPhase`] trait documentation.
+#[derive(Debug, Copy, Clone, PartialEq, Eq, Hash)]
+pub struct Initialized;
+
+impl sealed::Sealed for Initialized {}
+impl BootstrapPhase for Initialized {
+    type RequestHandler = BoxedHandler;
+    type Config = <WithApps as BootstrapPhase>::Config;
+    type Apps = <WithApps as BootstrapPhase>::Apps;
+    type Router = <WithApps as BootstrapPhase>::Router;
+    type AuthBackend = Box<dyn AuthBackend>;
+    #[cfg(feature = "db")]
+    type Database = Option<Arc<Database>>;
+}
+
+/// A part of [`CotProjectTempNameToChange`] that contains the shared context
```

Comment:
That needs changed here and in other places

Reply:
ah fair lol